### PR TITLE
Support for agent static network test

### PR DESCRIPTION
**Pull request by bfournie**
Add support for testing the agent static network configuration. Currently only the IP for Node 0 can be configured but the framework is there to use static IPs for all nodes when that support is available.

### CODE REVIEW

**Comment by pawanpinjarkar** [comment]
Tested this PR and /lgtm

File: agent/03_agent_deployment.sh [hunk 1]
```diff
@@ -13,6 +13,13 @@ source $SCRIPTDIR/agent/common.sh
 
 early_deploy_validation
 
+function add_host_dns_entries() {
+    ip_list=$1
+
+    sudo virsh net-update ostestbm add dns-host  "<host ip='${ip_list[0]}'> <hostname>master-0-static</hostname> </host>"  --live --config
```
Comment (dhellmann) [inline agent/03_agent_deployment.sh:19]: Doesn't the network name depend on the cluster name? I think there's a variable in common.sh.

Comment (bfournie) [reply agent/03_agent_deployment.sh]: Yes changed to use variable for network name

File: agent/03_agent_deployment.sh [hunk 2]
```diff
@@ -13,6 +13,18 @@ source $SCRIPTDIR/agent/common.sh
 
 early_deploy_validation
 
+function add_host_dns_entries() {
+    ip_list=$1
+    num_ips=${#ip_list[@]}
+
+    for (( i=0; i<$num_ips; i++ ))
+    do
+        hostname=${FLEETING_HOSTNAME}${i}
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:22]: Especially if in future we're going to add more entries for the other nodes, then I guess it could be better to be coherent with the current dev-scripts naming conventions and use `MASTER_HOSTNAME_FORMAT` and `WORKER_HOSTNAME_FORMAT`

File: agent/03_agent_deployment.sh [hunk 3]
```diff
@@ -13,6 +13,18 @@ source $SCRIPTDIR/agent/common.sh
...
+        sudo virsh net-update ${BAREMETAL_NETWORK_NAME} add dns-host  "<host ip='${ip_list[i]}'> <hostname>${hostname}</hostname> </host>"  --live --config
+    done
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:24]: As per the current imlementation, the same hostname could have an entry in the dhcp as well. What about removing such entry form the dhcp section, in order to have a cleaner/simpler environment?

File: agent/03_agent_deployment.sh [hunk 4]
```diff
@@ -107,17 +119,70 @@ stringData:
...
+
+    # Get the generated mac addresses to set up nmstate
+    mac_list=($(sudo virsh net-dumpxml ${BAREMETAL_NETWORK_NAME} | grep -Po '(?<=host mac=).*' | cut -d' ' -f1))
+    ip_list=$2
+
+    add_host_dns_entries ${ip_list}
+
+    num_ips=${#ip_list[@]}
+
+    echo -n "" > ${FLEETING_MANIFESTS_PATH}/nmstateconfig.yaml
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:131]: Is this effectively required?

File: agent/03_agent_deployment.sh [hunk 5]
```diff
@@ -107,17 +119,70 @@ stringData:
...
+    interfaces:
+      - name: eth0
+        type: ethernet
+        state: up
+        mac-address: ${mac_list[i]}
+        ipv4:
+          enabled: true
+          address:
+            - ip: ${ip_list[i]}
+              prefix-length: 24
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:155]: Here you could use `CLUSTER_HOST_PREFIX`

File: agent/03_agent_deployment.sh [hunk 6]
```diff
@@ -134,10 +199,34 @@ function attach_fleeting_iso() {
 
 write_pull_secret
 
-node0_ip=$(nth_ip $EXTERNAL_SUBNET_V4 20)
+# needed for assisted-service to run nmstatectl
+# This is temporary and will go away when https://github.com/nmstate/nmstate is used
+sudo yum install -y nmstate
+
+if [[ "$FLEETING_STATIC_IP_NODE0_ONLY" = "true" ]]; then
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:206]: Is this config variable required as a transient one?

File: agent/03_agent_deployment.sh [hunk 7]
```diff
@@ -107,17 +119,70 @@ stringData:
   .dockerconfigjson: '${pull_secret}'
 
 EOF
+
+    # Get the generated mac addresses to set up nmstate
+    mac_list=($(sudo virsh net-dumpxml ${BAREMETAL_NETWORK_NAME} | grep -Po '(?<=host mac=).*' | cut -d' ' -f1))
+    ip_list=$2
+
+    add_host_dns_entries ${ip_list}
```
[the remark this answers is not in the document]
Comment (bfournie) [reply agent/03_agent_deployment.sh:127]: Sure, moved to after call to generate_fleeting_iso()

File: agent/03_agent_deployment.sh [hunk 8]
```diff
@@ -107,17 +119,70 @@ stringData:
   .dockerconfigjson: '${pull_secret}'
 
 EOF
+
+    # Get the generated mac addresses to set up nmstate
+    mac_list=($(sudo virsh net-dumpxml ${BAREMETAL_NETWORK_NAME} | grep -Po '(?<=host mac=).*' | cut -d' ' -f1))
+    ip_list=$2
+
```
[the remark this answers is not in the document]
Comment (bfournie) [reply agent/03_agent_deployment.sh:126]: Good idea to tie the macs to the master/worker names. Made change per suggestion.

File: agent/03_agent_deployment.sh [hunk 9]
```diff
@@ -13,6 +13,18 @@ source $SCRIPTDIR/agent/common.sh
 
 early_deploy_validation
 
+function add_host_dns_entries() {
+    ip_list=$1
+    num_ips=${#ip_list[@]}
+
+    for (( i=0; i<$num_ips; i++ ))
```
[the remark this answers is not in the document]
Comment (bfournie) [reply agent/03_agent_deployment.sh:20]: I do like that syntax better but in this case the index is being used to get the hostname.

File: agent/03_agent_deployment.sh (hunk 2 above)
Comment (bfournie) [reply agent/03_agent_deployment.sh:22]: OK, changed to use the MASTER_HOSTNAME_FORMAT and WORKER_HOSTNAME_FORMAT instead of redefining it

File: agent/03_agent_deployment.sh (hunk 3 above)
Comment (bfournie) [reply agent/03_agent_deployment.sh:24]: I've run into some intermittent issues when testing this since it requires restarting the network after that change. I don't want to make this fix now and introduce instability but will take it as a follow-on.

File: agent/03_agent_deployment.sh (hunk 4 above)
Comment (bfournie) [reply agent/03_agent_deployment.sh:131]: Since the cat now appends to the file I wanted to make sure the file doesn't or is empty, but it looks like 'make clean' removes all of the manifest files so yes I can remove this line.

File: agent/03_agent_deployment.sh (hunk 5 above)
Comment (bfournie) [reply agent/03_agent_deployment.sh:155]: Good idea, changed to use that.

File: agent/03_agent_deployment.sh (hunk 6 above)
Comment (bfournie) [reply agent/03_agent_deployment.sh:206]: We need this now as the support for static ips on all states hasn't merged yet but I think it could also be used in the future as a case with static IP on node0 and dhcp addresses on all other hosts is a valid use case.

File: agent/03_agent_deployment.sh [hunk 10]
```diff
@@ -107,17 +134,64 @@ stringData:
...
 
 function generate_fleeting_iso() {
     export REPO_PATH=${WORKING_DIR}
+    ip_list=$1
+
     sync_repo_and_patch fleeting https://github.com/openshift-agent-team/fleeting ${FLEETING_PR}
 
-    generate_fleeting_manifests ${FLEETING_PATH}
-    
+    generate_fleeting_manifests ${FLEETING_PATH} $ip_list $2 $3
```
Comment (rwsu) [inline agent/03_agent_deployment.sh:192]: I'm having trouble testing it. Seeing this error:
`./agent/03_agent_deployment.sh: line 192: $2: unbound variable`

Comment (bfournie) [reply agent/03_agent_deployment.sh]: Thanks @rwsu, that should be fixed with latest update.

File: agent/03_agent_deployment.sh [hunk 11]
```diff
@@ -134,10 +208,40 @@ function attach_fleeting_iso() {
...
+    ip=${base_ip}+${n}
+    ip_list+=($(nth_ip ${external_subnet} ${ip}))
+done
+
 set_api_and_ingress_vip
 
-generate_fleeting_iso ${node0_ip}
+mac_list=()
+set_mac_addresses master $NUM_MASTERS
+set_mac_addresses worker $NUM_WORKERS
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:240]: I think there could be some space to improve the management of the static ips and mac lists. The two lists are synchronized, as their nth index refers to the same host, but there's no clue about that in the name. Also, their creation is performed in separate places (and with different approaches), and I think that this could make easier to introduce some errors in future.
I'd suggest to:
1) Use a more consistent namings for them, ie something like global vars `FLEETING_NODES_IPS` and `FLEETING_NODES_MACS` (btw simplify also the parameters passing)
2) Have just one function for generating both of them in the first level of the script
3) Some strong assumption about ordering are made elsewhere (ie first masters and then workers). That's reasonable when dealing with a single list, but please let's document it to avoid missing it in future

File: agent/03_agent_deployment.sh [hunk 12]
```diff
@@ -134,10 +208,40 @@ function attach_fleeting_iso() {
...
+
+if [[ "$IP_STACK" = "v4" ]]; then
+    external_subnet=$EXTERNAL_SUBNET_V4
+else
+    external_subnet=$EXTERNAL_SUBNET_V6
+fi
+
+# Use an IP address outside of the range that is set for DHCP, just for sanity checks.
+# It shouldn't matter as the dhcp addresses will not be picked up when using static IPs
+base_ip=40
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:229]: Current dhcp range ends up to 60, just to avoid any potentital conflict it's better to stay out of it

Comment (andfasano) [reply agent/03_agent_deployment.sh:229]: Additionally, see also here: https://github.com/metal3-io/metal3-dev-env/blob/681b4d26ab5f90b0d08679df36b82ebb74eca993/lib/network.sh#L133-L134

File: agent/03_agent_deployment.sh (hunk 11 above)
Comment (bfournie) [reply agent/03_agent_deployment.sh]: Good points @andfasano. I changed to use globals and reworked it to use one function to set everything up to ensure consistency.

File: agent/03_agent_deployment.sh (hunk 12 above)
Comment (bfournie) [reply agent/03_agent_deployment.sh]: Changed to start IPs at 80 to avoid any overlap with DHCP.

File: agent/03_agent_deployment.sh [hunk 13]
```diff
@@ -107,17 +134,64 @@ stringData:
...
+    num_ips=${#ip_list[@]}
+
+    # Create a yaml for each host in nmstateconfig.yaml
+    for (( i=0; i<$num_ips; i++ ))
+    do
+        cat >> "${FLEETING_MANIFESTS_PATH}/nmstateconfig.yaml" << EOF
+apiVersion: agent-install.openshift.io/v1beta1
+kind: NMStateConfig
+metadata:
+  name: mynmstateconfig
```
[the remark this answers is not in the document]
Comment (bfournie) [reply agent/03_agent_deployment.sh]: Thanks @rwsu, yes good point. I changed to use the hostname in the manifest that is generated at the same time as the IPs and Macs (see comment above).

File: agent/03_agent_deployment.sh [hunk 14]
```diff
@@ -107,17 +150,61 @@ stringData:
...
 }
 
 function generate_fleeting_iso() {
     export REPO_PATH=${WORKING_DIR}
+
     sync_repo_and_patch fleeting https://github.com/openshift-agent-team/fleeting ${FLEETING_PR}
 
-    generate_fleeting_manifests ${FLEETING_PATH}
-    
+    generate_fleeting_manifests ${FLEETING_PATH} ${FLEETING_NODES_IPS}
```
Comment (andfasano) [inline agent/03_agent_deployment.sh:205]: nit: this could be simplified, but we could do it into another pr

Comment (bfournie) [reply agent/03_agent_deployment.sh]: Good catch, updated.

**Comment by andfasano** [comment]
ipv6 failures are not related to this patch

/override ci/prow/e2e-metal-ipi-ovn-ipv6
/override ci/prow/e2e-metal-ipi-serial-ovn-ipv6
